src/composer.ts

```diff
@@ -908,24 +944,65 @@ export class Composer<C extends Context> implements MiddlewareObj<C> {
             next: NextFunction,
         ) => MaybePromise<unknown>,
         ...middleware: Array<Middleware<C>>
-    ) {
-        const composer = new Composer<C>(...middleware);
-        const bound = flatten(composer);
-        this.use(async (ctx, next) => {
+    ): Composer<C> {
+        return this.boundary(async (ctx, subtree) => {
             let nextCalled = false;
-            const cont = () => ((nextCalled = true), Promise.resolve());
+            const next = () => ((nextCalled = true), Promise.resolve());
             try {
-                await bound(ctx, cont);
-            } catch (err) {
-                nextCalled = false;
-                await errorHandler(new BotError<C>(err, ctx), cont);
+                return await subtree();
+            } catch (e) {
+                await errorHandler(new BotError(e, ctx), next);
+                return nextCalled;
             }
-            if (nextCalled) await next();
-        });
-        return composer;
+        }, ...middleware);
     }
 }
 
+/**
+ * > This is an advanced type of grammY.
+ *
+ * Wraps a middleware subtree, in order to control the behavior of a boundary.
+ *
+ * `await subtree()` calls the subtree. The returned boolean indicates
+ * whether the subtree attempted to pass control outside,
+ * by calling `next()` in the last middleware inside.
+ *
+ * Return `true` to actually pass control outside after the controller settles,
+ * or `false` not to. When in doubt, return what `await subtree()` did.
+ *
+ * The following example will reach the points A-F alphabetically.
+ *
+ * ```ts
+ * // wrap a subtree of middleware with custom logic
+ * const sub = bot.surround(
+ *   // pass a boundary controller, wrapping the subtree
+ *   async (ctx, subtree) => {
+ *     // A
+ *     await subtree() // -- returns false because C didn't call next()
+ *     // E
+ *   },
```

Comment:
Should there be a `return` in this doc to conform to the expected type?
```suggestion
 *   async (ctx, subtree) => {
 *     // A
 *     const nextCalled = await subtree() // -- returns false because C didn't call next()
 *     // E
 *     return nextCalled
 *   },
```